odoo/addons/base/res/res_users.py

```diff
@@ -510,6 +514,36 @@ def check(cls, db, uid, passwd):
         finally:
             cr.close()
 
+    def _get_session_token_fields(self):
+        token_fields = set(USER_PRIVATE_FIELDS + ['active', 'password_crypt'])
+        return set(self.fields_get_keys()).intersection(token_fields)
+
+    @tools.ormcache('sid')
+    def _compute_session_token(self, sid):
+        """ Compute a session token given a session id and a user id """
+        # retrieve user's password crypt
+        session_fields = ', '.join(sorted(self._get_session_token_fields()))
+        self.env.cr.execute("""SELECT %s, (SELECT value FROM ir_config_parameter WHERE key='database.secret')
+                                FROM res_users
+                                WHERE id=%%s""" % (session_fields), (self.id,))
+        if self.env.cr.rowcount != 1:
+            self._invalidate_session_cache()
+            return False
+        # 'stringify' the fields to use as the key
+        data_fields = str(self.env.cr.fetchone())
+        # generate hmac key
+        key = (u'%s-%s' % (self.id, data_fields)).encode('utf-8')
+        # hmac the session id
+        data = sid.encode('utf-8')
+        h = hmac.new(key, data, sha1)
```

Comment:
Not technically necessary but given task 31589 (phase out sha1) maybe use a sha2 hash instead?